Marlin/src/feature/bltouch.cpp

```diff
@@ -56,41 +46,138 @@ bool BLTouch::triggered() {
   );
 }
 
-bool BLTouch::set_deployed(const bool in_deploy) {
-  if (in_deploy && triggered()) {     // If BLTouch says it's triggered
-    reset();                          //  try to reset it.
-    _deploy(); _stow();               // Deploy and stow to clear the "triggered" condition.
-    safe_delay(1500);                 // Wait for internal self-test to complete.
-                                      //  (Measured completion time was 0.65 seconds
-                                      //   after reset, deploy, and stow sequence)
-    if (triggered()) {                // If it still claims to be triggered...
-      SERIAL_ERROR_MSG(MSG_STOP_BLTOUCH);
-      stop();                         // punt!
-      return true;
-    }
+void BLTouch::init() {
+  // This is called by marlin.cpp on initialization
+  // SET_5V_MODE (if enabled). OD_MODE is the default on power on.
+  // This mode will stay active until manual SET_OD_MODE or power cycle
+  #if ENABLED(BLTOUCH_FORCE_5V_MODE)         
+    _set_5V_mode();                          // Set 5V mode if explicitely demanded (V3 upwards)
+  #endif
+  clear();
+  // There really should be no alarm outstanding now, and no triggered condition. But if there is,
+  // there is no need to worry people here on init right at the start of the printer.
+}
+
+void BLTouch::clear() {
+  _reset();    // RESET or RESET_SW will clear an alarm condition but...
+               // ...it will not clear a triggered condition in SW mode when the pin is currently up
+               // ANTClabs <-- CODE ERROR
+  _stow();     // STOW will pull up the pin and clear any triggered condition unless it fails, don't care
+  _deploy();   // DEPLOY to test the probe. Could fail, don't care
+  _stow();     // STOW to be ready for meaningful work. Could fail, don't care
+}
+
+bool BLTouch::command(const BLTCommand cmd, millis_t ms) {
+  #if ENABLED(BLTOUCH_DEBUG_MSGS)
+    SERIAL_ECHOLNPAIR("BLTouch Command :", cmd);
+  #endif
+  MOVE_SERVO(Z_PROBE_SERVO_NR, cmd);
+  // If BLTOUCH_DELAY is higher than the desired ms delay, it will override
+  if (BLTOUCH_DELAY > ms) {
+    ms = BLTOUCH_DELAY;
+  }
+  // Don't wait longer than necessary if an ALARM pops up. Check trigger
+  // in increments of 50ms to catch the ALARM right away.
```

Comment:
`safe_delay()`:
```
void safe_delay(millis_t ms) {
  while (ms > 50) {
    ms -= 50;
    delay(50);
    thermalManager.manage_heater();
  }
  delay(ms);
  thermalManager.manage_heater(); // This keeps us safe if too many small safe_delay() calls are made
}
```
Why might this be off by 49ms? It does a possible rest value outside the loop at the end?

Reply:
> There is a good chance this will be picked up by this at some point.

I though that unless something is very wrong, there should never be a trigger while in this loop. If this is a wait after a DEPLOY, movement only begins after the wait is over. Similar logic applies to the other commands.

> Should have a flag for triggered seen and only fault out of true twice in a row.

But that is a really good idea to not get out of the loop when there is noise.

And you are right: In no case do we want to get into a situation where we might issue the next command before the wait for the previous command is over.

I think it would be best to scrap this exiting of the loop altogether. I kind of though it would speed up things - but we are talking about reducing (for example) a 750ms STOW wait to 350ms, who cares actually. How often do we probe, anyway.

So I will remove the two lines and let's let the wait do its thing without any pre-exit.

Reply:
> Could do while(ms > 0) then safe_delay MIN(ms, 50));

Yes very nice idea for the original `safe_delay` by the way, but how does the code bytes overhead of `MIN` compare to just doing the rest after the end of the loop?

Reply:
The 'new idea' will break things...    If safe_delay() is called with a small amount of time...   say 25ms, no delay will be performed.

Reply:
Yup. Of course you are right. like switching those two lines, lemme see:
```
void safe_delay(millis_t ms) {
  while (ms > 50) {
    delay MIN(ms, 50))
    ms -= 50;
    thermalManager.manage_heater();
  }
}
```
I was too quick off the mark but @InsanityAutomation idea goes along those lines I think.

Reply:
Yeah!    That code is super easy to KNOW the right thing is going to happen.   I much prefer that form.    And incidentally...  That is more 'optimum' too.   That will generate less code than what is there right now.

Reply:
I have checked:

The "new idea" looks better in the c code. As I feared: It is longer on AVR machine code. It is slightly longer on 32 bit code.  The original code knew what he was doing. 

Do we care about 10-20 bytes of machine code in this specific case? In general, we do.

Reply:
10 or 20 bytes isn't a big deal to make sure the problem is fixed...   With regard to MIN() being expensive...    In this case we know the types going into the macro.   If we side step the MIN() macro and just use a ternary operator, I would think this ends up as small as possible:
```
void safe_delay(millis_t ms) {
  while (ms > 0) {
    delay( ms<50? ms : 50);
    ms -= 50;
    thermalManager.manage_heater();
  }
}
```

Reply:
Yup. The original code knew what he was doing. `ms<50? ms : 50` in machine code is loads for the ms var, multiple compares and 2 branches. The original one is shorter in machine code, believe me. But less shorter on 32 bit machines. AVR - it's hell.

If I were a CPU I would go berserk anyway, first the guy is asking `while (ms > 50)` god knows how many times, and then he wants to know `ms < 50` that many times. Bonkers, right?

Reply:
> to make sure the problem is fixed

But there is actually no problem